Replace check+cast pairs with just casts

amomchilov opened the pull request:
### Motivation

This pattern is clunky and error-prone:

```c++
if (PM_NODE_TYPE_P(node, PM_BLOCK_PARAMETERS_NODE)) {
    auto *blockParam = down_cast<pm_block_parameter_node>(node);
    ...
}
```

Did you catch the extra `S`?

This PR reworks the semantics of the `down_cast()` helper to do both of these in one operation, similar to `parser::cast_node()` and `ast::cast_tree()`. It does a bit of a renaming dance:

1. Rename `down_cast()` to `down_cast_nonnull()`, whose name better describes its existing semantics
   * Enforces non-null and correct type
2. Introduce a new function as `down_cast()`
   * Returns null on null, or incorrect type
3. Add a similar `isa_node()` helper for consistency (and to dedupe some explicit null checks)
   * In the vast majority of calls, the `null` check can be skipped. So there's `isa_node_nonnull()`, to only add the null checks where needed.

### Test plan

Pure refactor, covered by existing desugar tests.

amomchilov committed:
Extract `isa_node()` helper

jez reviewed (changes requested):
What do you think about making this work the way that all other `cast_*` functions work in Sorbet: if given `nullptr`, they return `nullptr`:

- [cast_tree](https://github.com/sorbet/sorbet/blob/master/ast/Trees.h#L216-L226)
- [cast_node](https://github.com/sorbet/sorbet/blob/7232081ea6f1bf89eef21d638fd193c4ff85e118/common/common.h#L113-L115)
- [cast_type](https://github.com/sorbet/sorbet/blob/7232081ea6f1bf89eef21d638fd193c4ff85e118/core/TypePtr.h#L250-L260)

This is also the way that C++'s [dynamic_cast](http://www.en.cppreference.com/w/cpp/language/dynamic_cast.html#:~:text=a%20null%20pointer%20value%20if%20expression%20is%20a%20null%20pointer%20value%2C%20or) keyword works.

As in, every other downcasting operation in Sorbet already works like `try_down_cast` in the `nullptr` case.

amomchilov committed:
Rename `down_cast` → `down_cast_nonnull`

amomchilov committed:
Replace check+casts with just casts

Add a nullable `down_cast` helper that returns `nullptr` if the node is
null or not of the expected type. This matches the convention of Sorbet's
other downcasting helpers like `cast_tree`, `cast_node`, `cast_type`, and
C++'s `dynamic_cast`.

This lets us collapse the common `PM_NODE_TYPE_P` + `down_cast_nonnull`
pattern:

    if (PM_NODE_TYPE_P(node, PM_FOO_NODE)) {
        auto fooNode = down_cast_nonnull<pm_foo_node>(node);
        ...
    }

into:

    if (auto *fooNode = down_cast<pm_foo_node>(node)) {
        ...
    }

amomchilov committed:
Extract `isa_node_nullable()` helper

amomchilov committed:
Delete redundant checks

amomchilov committed:
Rename `stmtsNode` to `body`
